Hypershift live migration (#3478)
ovn-org/ovn-kubernetes: docs/live-migration.md

the VM should appear as the following with `kubectl get vmi fedora`

```bash
kubectl get vmi fedora
NAME     AGE     PHASE     IP            NODENAME      READY
fedora   9m42s   Running   10.244.2.26   ovn-worker3   True
```

Login and check that the VM has receive a proper address `virtctl console fedora`
```bash
[fedora@fedora ~]ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc fq_codel state UP group default qlen 1000
    link/ether 0a:58:0a:f4:02:1a brd ff:ff:ff:ff:ff:ff
    altname enp1s0
    inet 10.244.2.26/24 brd 10.244.2.255 scope global dynamic noprefixroute eth0
       valid_lft 3412sec preferred_lft 3412sec
    inet6 fe80::32d2:10d4:f5ed:3064/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
```

Also we can check the neighbours cache to verify it later on
```bash

> [fedora@fedora ~]arp -a

Reviewer: that's exactly why I suggested using `ip n` instead. 

From what I understand, the `net-tools` package - from which the arp command originates - is no longer actively developed, and most of those tools have been deprecated in favor of their ip equivalents.

_gateway (169.254.1.1) at 0a:58:a9:fe:01:01 [ether] on eth0
```

Something that may be unexpected is that the default gw is a link local address, that's because
the live migration feature is implemented using ARP proxy, the last route is
needed since the link local address subnet is not part of any interface, that
route is automatically created by dhcp client.

```bash
[fedora@fedora ~]ip route
default via 169.254.1.1 dev eth0 proto dhcp src 10.244.2.26 metric 100
10.244.2.0/24 dev eth0 proto kernel scope link src 10.244.2.26 metric 100
169.254.1.1 dev eth0 proto dhcp scope link src 10.244.2.26 metric 100
```

Then a live migration can be initialized with `virtctl migrate fedora` and wait
at the vmim resource
```bash
virtctl migrate fedora
VM fedora was scheduled to migrate
kubectl get vmim -A -o yaml
  status:
    migrationState:
      completed: true
```

> After this all the network configuration is exactly the same including the
> gw neightbours cache

Reviewer: ```suggestion
After migration, the network configuration is the same - including the GW neighbor cache.
```

default dns service `kube-system/kube-dns` as the name server, but this can be
overriden with the following command line options:
- dns-service-namespace
- dns-service-name

# Configuring dual stack guest images
For dual stack ovn-kubernetes is delivering the IPv6 address to guest VMs using
DHCPv6 but the IPv6 default gateway has to be configured manually, this address
is always the same `d7b:6b4d:7b25:d22f::1` so the virtual machine configuration
is stable. Also both ipv4 and ipv6 configurations has to be activated.

NOTE: The IPv6 autoconf/SLAAC is not supported at ovn-k live-migration

For fedora cloud-init can be used to activate dual stack:
```yaml
 - cloudInitNoCloud:
     networkData: |
       version: 2
       ethernets:
         eth0:
           dhcp4: true
           dhcp6: true
     userData: |-
       #cloud-config
       # The default username is: fedora
       password: fedora
       chpasswd: { expire: False }
       runcmd:
         - nmcli c m "cloud-init eth0" ipv6.method dhcp
         - nmcli c m "cloud-init eth0" +ipv6.routes "d7b:6b4d:7b25:d22f::1"
         - nmcli c m "cloud-init eth0" +ipv6.routes "::/0 d7b:6b4d:7b25:d22f::1"

>          - nmcli c reload "cloud-init eth0"

Reviewer: wait so you're saying I can't add a route using netplan v2 ? 

I also don't follow your second point: 
> also dhcpv6=true activates SLAAC and that breaks this

I literally see `dhcp6: true` in the netplan configuration above.

# Live Migration

## Introduction

The Live Migration feature allows [kubevirt](kubevirt.io) virtual machines to be [live migrated](https://kubevirt.io/user-guide/operations/live_migration/)
while keeping the established TCP connections alive, and preserving the VM IP configuration.
These two requirements provide seamless live-migration of a KubeVirt VM using OVN-Kubernetes
cluster default network.

## Requirements

> - KubeVirt >= v1.0.0

Reviewer: could you refresh my mind into this ? What did you add to kubevirt in this particular release?

Author: It's at the PR description:

future kubevirt 1.0:
- https://github.com/kubevirt/kubevirt/pull/9290
- https://github.com/kubevirt/kubevirt/pull/9603

- DHCP aware guest image (fedora family is well tested, https://quay.io/organization/containerdisks)

# Limitations

- Only KubeVirt VMs with bridge binding pod network are supported
- Single stack IPv6 is not supported

> - DualSack do not configure routes for IPv6 over DHCP/autoconf

Reviewer: ```suggestion
- DualSack does not configure routes for IPv6 over DHCP/autoconf
```

Reviewer: further question: this means the routes must manually configured routes ?

Author: Yes, the router advertisment support at ovn-k will advertise LRP but we need the arp proxy default gateway wich is not advertised, but since the default gw is stable across nodes we can configure it at general ignition at hypershift or cloud-init at other envs.

- SRIOV is not supported

## Example: live migrating a fedora guest image

Install KubeVirt following the guide [here](https://kubevirt.io/user-guide/operations/installation/)

> Create a fedora virtual machine with the annotations `kubevirt.io/allow-pod-bridge-network-live-migration` and `k8s.ovn.org/skip-ip-configuration-on-cni`:

Reviewer: didn't you remove the `skip-ip-configuration` annotation ?

Author: Done
